src/graph.mojo

```diff
@@ -0,0 +1,163 @@
+    var neighbours: InlineArray[Int, 256]
+
+    # The mask of the node.
+    var bit_mask: BitMask
+
+    fn __init__(inout self, owned bit_mask: BitMask, owned value: DataType):
+        """Initializes the node with the given mask and value.
+
+        Args:
+            bit_mask: The bit mask of the node.
+            value:    The value stored in the node.
+        """
+        self.value = value
+        self.neighbours = InlineArray[Int, 256](Self.null_index)
+        self.bit_mask = bit_mask
+
+
+struct BitMaskGraph[DataType: AnyTrivialRegType, //, null_value: DataType]:
+    """A graph where each node is identified by a BitMask.
+
+    The graph is intended to be used for fast lookup of data
+    identified by bitmasks based on a known starting point
+    differing only in a few bits from the sought-after node.
+
+    Nodes that differ by exactly one bit are connected via edges.
+    Edges are constructed upon first use.
+
+    Parameters:
+        DataType:   The type of the value stored in the nodes.
+        null_value: The place holder stored in nodes by default.
+    """
+
+    # The node index indicating a non-established link.
+    alias null_index = Node[DataType].null_index
+
+    # The list of nodes in the graph.
+    var _nodes: List[Node[DataType]]
+
+    # A mapping for random lookup of nodes by their mask.
+    # Used for slow lookup of nodes.
+    var _map: Dict[BitMask, Int]
+
+    fn __init__(inout self, owned first_value: DataType = Self.null_value):
+        """Initializes the graph.
+
+        Args:
+            first_value: The value stored in the first node,
+                         corresponding to an empty bitmask.
+        """
+        self._nodes = List[Node[DataType]]()
+        self._map = Dict[BitMask, Int]()
+        _ = self.add_node(BitMask(), first_value)
+
+    @always_inline
+    fn add_node(
+        inout self,
+        owned node_mask: BitMask,
+        owned value: DataType = Self.null_value,
+    ) -> Int:
+        """Adds a node to the graph.
+
+        Args:
+            node_mask: The mask of the node.
+            value:     The value stored in the node.
+
+        Returns:
+            The index of the added node.
+        """
+        self._map[node_mask] = len(self._nodes)
+        self._nodes.append(Node(node_mask, value))
+        return len(self._nodes) - 1
+
+    @always_inline
+    fn create_link(
+        inout self, from_node_index: Int, changed_bit: BitMask.IndexType
+    ) -> Int:
+        """Creates a link between two nodes.
+
+        Args:
+            from_node_index: The index of the node from which the link is created.
+            changed_bit:     The index of the bit that differs between the nodes.
+
+        Returns:
+            The index of the node to which the link is created.
+        """
+        new_mask = self._nodes[from_node_index].bit_mask
+        new_mask.flip(changed_bit)
+        optional_to_index = self._map.get(new_mask)
```

Comment:
I guess you plan to call this method only if there is not a link yet (which is what the method name implies)? Otherwise, you should check that here. In Arche, there is no separate type for the graph, and I do this all in one method, where it looks like this:

```go
if next, ok := curr.neighbors.Get(id.id); ok {
    curr = next
} else {
    next, _ := w.findOrCreateArchetypeSlow(mask, relation, hasRelation)
    next.neighbors.Set(id.id, curr)
    curr.neighbors.Set(id.id, next)
    curr = next
}
```